pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/TopicSchema.java

```diff
@@ -196,6 +197,9 @@ private static boolean isProtobufClass(Class<?> pojoClazz) {
         String schemaTypeOrClassName = conf.getSchemaType();
         if (StringUtils.isEmpty(schemaTypeOrClassName) || DEFAULT_SERDE.equals(schemaTypeOrClassName)) {
             // No preferred schema was provided, auto-discover schema or fallback to defaults
+            if (!input && clazz.equals(GenericRecord.class)) {
+                return new AutoProduceBytesSchema();
+            }
```

Comment:
It seems integration test don't cover here, it also use `record.getSchema()`. What is the integration test testing? I did not understand here.

Reply:
The integration test without this change does not work.

Reply:
you start a source, but source don't getSchema by this method. I don't understand where use this change?

Reply:
@congbobo184 
This method is called by `PulsarSink` in `initializeSchema() `method via `TopicSchema#getSchema`
The PulsarSink is the entity that actually write to Pulsar.
When you create a Source Pulsar IO creates a PulsarSink to write to the destination topic. 

When you start the Source you are going to use this special `AutoProduceBytesSchema` (it was pre-existing, I did not add it).

Initially the Source does not enforce a Schema on the topic (we achieve this with `AutoProduceBytesSchema`).

When the Source passes a Record to the Pulsar runtime, the PulsarSink picks up the Schema (using Record#getSchema) and sets the schema properly.
Therefore when the schema changes (in a compatible way) the runtime automatically updates the schema.

In fact when the Source starts you cannot know the schema, because the schema is generated dynamically by the Source itself.

Reply:
Thank you for your detailed reply, I understand here.   :)